File: BinaryTrees/LowestCommonAncestor/lca.py
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None


class Solution:
    def checkRootExists(self,root,value):
        if not root: return False
        if root.data == value: return True
        return self.checkRootExists(root.left,value) or self.checkRootExists(root.right,value)
    def lca(self, root, n1, n2):
        # code here
        if not root: return None
        elif root.data == n1 or root.data == n2: return root
        elif ((self.checkRootExists(root.left,n1) and self.checkRootExists(root.right,n2)) or 
        (self.checkRootExists(root.left,n2) and self.checkRootExists(root.right,n1))): return root
        elif (self.checkRootExists(root.left,n1) and not self.checkRootExists(root.right,n2)): return self.lca(root.left,n1,n2)
        else: return self.lca(root.right,n1,n2)

File: BinaryTrees/LowestCommonAncestor/lca.cpp
#include <bits/stdc++.h>
using namespace std;

// Definition for Node
class Node {
public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = right = nullptr;
    }
};

class Solution {
public:
    bool checkRootExists(Node* root, int value) {
        if (!root)
            return false;

        if (root->data == value)
            return true;

        return checkRootExists(root->left, value) ||
               checkRootExists(root->right, value);
    }

    Node* lca(Node* root, int n1, int n2) {
        if (!root)
            return nullptr;

        if (root->data == n1 || root->data == n2)
            return root;

        if ((checkRootExists(root->left, n1) &&
             checkRootExists(root->right, n2)) ||
            (checkRootExists(root->left, n2) &&
             checkRootExists(root->right, n1)))
            return root;

        else if (checkRootExists(root->left, n1) &&
                 !checkRootExists(root->right, n2))
            return lca(root->left, n1, n2);

        else
            return lca(root->right, n1, n2);
    }
};